### ISSUE TITLE
Nullable Reference Types: A warning should be reported when dereferencing a field/property before it is assigned

### ISSUE THREAD

**Issue by PathogenDavid** [dotnet/roslyn#27016]
**Version Used**: 05/14/18 Nullable Reference Types Preview (`csc` reports `2.8.0.62830 (e595ee27)`) with Visual Studio 15.7.1

**Demonstration Code**:

```csharp
public class TestClass
{
    private string x;

    public TestClass()
    {
        Console.WriteLine(x.ToString()); // No warning, CS8602 is expected
        x = "Hello, world!";
    }
}
```

**Expected Behavior**:

A warning is reported when `x` is dereferenced.

**Actual Behavior**:

There is no compile-time warning and a runtime exception occurs when the code is executed.

----

**Additional info**:

The current behavior is the same with a property.

**Comment by jaredpar** [comment]
Thanks for reporting. At this point the behavior is by design. The field is declared as non-nullable hence the compiler treats it that way. In this example yes it's easy but in other examples, base constructors, chained constructors, the analysis can't be done as easily.

**Comment by jaredpar** [comment]
@agocke for the greater ctor tracking issues

**Comment by YairHalberstadt** [comment]
I find this a really common source of bugs.

If a constructor does not call a base constructor, or a this constructor, I believe all fields will always be null.

So a simple rule would be that in such a case, all fields start off with a nullable state of maybe null, no matter their annotation, unless they are initialized at the declaration site.

I think this shouldn't apply to properties - just to fields.
